File: Java/DynamicProgramming/DP52 - Evaluate Boolean Expression to True.java
package DynamicProgramming;


class RecursiveSolutionDP52 {
    /**
     * Time complexity is exponential and space complexity is O(n).
     */

    private static Integer solve(String expression, Integer i, Integer j, Boolean evaluateTo) {
        if (i > j) {
            return 0;
        }
        if (i.equals(j)) {
            if (evaluateTo.equals(true)) {
                return expression.charAt(i) == 'T' ? 1 : 0;
            } else {
                return expression.charAt(i) == 'F' ? 1 : 0;
            }
        }
        int numWays = 0;
        for (int index = i + 1; index <= j - 1; index += 1) {
            Integer leftTrue = solve(expression, i, index - 1, true);
            Integer leftFalse = solve(expression, i, index - 1, false);
            Integer rightTrue = solve(expression, index + 1, j, true);
            Integer rightFalse = solve(expression, index + 1, j, false);
            char operator = expression.charAt(index);
            if (operator == '&') {
                if (evaluateTo.equals(true)) {
                    numWays += (leftTrue * rightTrue);
                } else {
                    numWays += (leftTrue * rightFalse + rightTrue * leftFalse + leftFalse * rightFalse);
                }
            } else if (operator == '|') {
                if (evaluateTo.equals(true)) {
                    numWays += (leftTrue * rightFalse + rightTrue * leftFalse + leftTrue * rightTrue);
                } else {
                    numWays += (leftFalse * rightFalse);
                }
            } else {
                if (evaluateTo.equals(true)) {
                    numWays += (leftTrue * rightFalse + rightTrue * leftFalse);
                } else {
                    numWays += (leftTrue * rightTrue + leftFalse * rightFalse);
                }
            }
        }
        return numWays;
    }

    public static Integer getNumWaysToEvaluateTo(String expression) {
        int n = expression.length();
        return solve(expression, 0, n - 1, true);
    }

    public static void main(String[] args) {
        System.out.println(getNumWaysToEvaluateTo("F|T^F"));
        System.out.println(getNumWaysToEvaluateTo("T|T&F"));
        System.out.println(getNumWaysToEvaluateTo("T^T^F"));
        System.out.println(getNumWaysToEvaluateTo("T|T&F^T"));
        System.out.println(getNumWaysToEvaluateTo("T^F|F"));
    }
}

File: Python/Practice Set 3/Dynamic Programming/DP52 - Evaluate Boolean Expression to True.py
def recursive():
    def solve(expression, i, j, evaluate_to):
        if i > j:
            return 0
        if i == j:
            if evaluate_to:
                return 1 if expression[i] == 'T' else 0
            return 1 if expression[i] == 'F' else 0
        num_ways = 0
        for index in range(i + 1, j, 2):
            operator = expression[index]
            left_true = solve(expression, i, index - 1, True)
            left_false = solve(expression, i, index - 1, False)
            right_true = solve(expression, index + 1, j, True)
            right_false = solve(expression, index + 1, j, False)
            if operator == "&":
                if evaluate_to:
                    num_ways += (left_true * right_true)
                else:
                    num_ways += (left_true * right_false + right_true * left_false + left_false * right_false)
            elif operator == "|":
                if evaluate_to:
                    num_ways += (left_true * right_true + left_true * right_false + left_false * right_true)
                else:
                    num_ways += (left_false * right_false)
            else:
                if evaluate_to:
                    num_ways += (left_true * right_false + right_true * left_false)
                else:
                    num_ways += (left_true * right_true + left_false * right_false)
        return num_ways

    def evaluate(expression):
        n = len(expression)
        return solve(expression, 0, n - 1, True)

    print(evaluate("F|T^F"))
    print(evaluate("T|T&F"))
    print(evaluate("T^T^F"))
    print(evaluate("T|T&F^T"))
    print(evaluate("T^F|F"))


recursive()
print()
